### ISSUE TITLE
How to make a composite genotype of permutation and constant sum constrained part?

### ISSUE THREAD

**Issue by macabrus** [jenetics/jenetics#838]
Hello,

First I'd like to say this is an amazing library and I am using it as a building block in preparation for my masters in CS!

I am sorry for asking this as an issue, but I am not sure how I'd go about defining a Codec for a Problem<> interface where I'd have a phenotypes to be generated of following shape:
<0...n permuted integers><m integers that add up to specified sum>
for example n=10, m=3 (which adds up to 10), a valid genotype would have single chromosome:
(2,1,3,5,4,6,10,8,9,7) | (5,4,1)
 first part   | second part

Just to note: I am solving mTSP problem where first part denotes permutation of all cities, and second part denotes how many cities in sequence i-th salesman has.

I tried to write it as Codec.of() for composite codec: first part as Codecs.ofPermutation as stated in documentation, but how to go about second part of chromosome? I haven't found something like Codecs.ofFixedSum(num_of_ints, total_sum) or something. I would be happy to implement this bit myself but I'd need guidance on what to override to make my own codec for generating valid phenotypes of population.

**Comment by jenetics** [comment]
I think you need to change the gene type. The following `Codec` should give you an idea how to do it.

```java
record Solution(int[] path, int[] lengths) {}

static Codec<Solution, DoubleGene> of(final int n, final int m) {
    return Codec.of(
        Genotype.of(
            // Encoding of (2,1,3,5,4,6,10,8,9,7)
            DoubleChromosome.of(DoubleRange.of(0, 1), n),
            // Encoding of (5,4,1)
            DoubleChromosome.of(DoubleRange.of(0, 1), m)
        ),
        gt -> {
            // Use the sorted indexes as path permutations.
            final int[] path = ProxySorter.sort(
                gt.get(0)
                    .as(DoubleChromosome.class)
                    .toArray()
            );

            // Calculate the 'length' array.
            final double[] values = gt.get(1)
                .as(DoubleChromosome.class)
                .toArray();

            final double sum = DoubleStream.of(values).sum();
            final int[] lengths = new int[m];
            for (int i = 0; i < m; ++i) {
                // Check for rounding errors, didn't test it.
                lengths[i] = (int)Math.round(values[i]/sum*n);
            }

            return new Solution(path, lengths);
        }
    );
}
```

**Comment by macabrus** [comment]
Thank you, I get the idea of your solution, I'll check for rounding errors and manually handle case if it is off in total sum at the end. Also, mTSP often comes with MIN and MAX constraints on each salesman's cities in tour for balancing load/ minimising total time of travel. I suppose it would be analogous to DoubleChromosome.of(DoubleRange.of( MIN / N, MAX / N)) where N = num of cities?

**Comment by jenetics** [comment]
@macabrus I have problems  understanding your questions. It might be totally clear to you, since you are currently more involved into this domain, but I'm not. If you can elaborate the problems in a greater detail, will help me to answer your questions. 

What will help me to help you:
* Describe the Problem you try to solve, as precise and narrow as possible.
* Your  proposed solution (preferable as code snippet).

Regards
Franz

**Comment by macabrus** [comment]
Hey, I'm sorry for not explaining in greater detail, I should not have assumed that everything is obvious. So basically, for this constrained type of chromosome, your proposed solution of using `DoubleChromosome.of(DoubleRange.of(0, 1), m)` has rounding error issues that the comment mentions, however I can't really think of a way to resolve rounding errors without introducing bias to specific array indices.

The idea I had in mind:
```
final double sum = DoubleStream.of(values).sum();
final int[] lengths = new int[m];
for (int i = 0; i < m; ++i) {
    // Check for rounding errors, didn't test it.
    lengths[i] = (int)Math.round(values[i]/sum*n);
}

int error = n - Stream.of(lengths).sum();
while (error != 0) {
    if(error < 0) {
        // how to choose which lengths[?] to increment without skewing probabilities and retaining 1:1 decodability?
        error++;
    } else {
        // same question but decrement
        error--;
    }
}
```

Other thought that comes to my mind is that these rounding errors are sufficiently rare not to nitpick about skewing probabilities? It probably has to do with infinite sets theory which I am not too familiar with.

If someone stumbles on this question again, I decided to go with alternative approach using **discrete convolutions** described [here](https://stackoverflow.com/questions/18448417/create-constrained-random-numbers/18448874#18448874).

Thanks for your time, this helped me understand Jenetics' concepts better. Shall I close this issue as to not bother you since it is not directly related to Jenetics but domain problem?

TL;DR It was not obvious to me how I would fix rounding errors.

**Comment by jenetics** [comment]
This was just a reminder to double-check the _length_ calculation. I don't think that it creates any bias during the search. Just the usual Index-Out-Of-Bounds error you have to check. I'm doing this for too long ;-)
